### DISCUSSION TITLE
Alpine Linux with QuickSync

### DISCUSSION THREAD

**Discussion by Blissingg** [OvenMediaLabs/OvenMediaEngine#1692] [Discover]
I Just wanted to document and share my process of getting OvenMediaEngine working on Alpine Linux with QuickSync in a container using Incus in case it helps anyone that wants to do similar.

### Background:
I recently got a new mini-pc with an Intel N100 for HomeAssistant and decided to leverage the intel CPU's ability to use Quicksync with OME

### Host OS:
Arch Linux using Incus as my container platform

### Container OS:
Alpine Linux 3.20

### Steps:

It wasn't too hard to get this working, it just required installing the dependencies of OME, quicksync and editing  three files  that were downloaded from the github repository to allow compiling of OME on Alpine Linux.

The three files edited were:

1. src/projects/base/ovcrypto/openssl/openssl_manager.cpp
2. `src/projects/monitoring/server_metrics.cpp`
3. `src/projects/main/signals.cpp`~

The edits done were the following lines

1. `COMP_zlib_cleanup();` Replaced with `OPENSSL_NO_COMP`
2. `malloc_trim(0);` Comment it out `//malloc_trim(0);`
3. `("Trim result: %d", malloc_trim(0));`Edited to `("Trim result: %d");`

To make it easier I created a script to download, compile and install OME if anyone wants it, it can be found at the following, replace JOBS=2 with the number of jobs your system can handle.

https://pastebin.com/x0hjNfmx

### QuickSync support:

If you want Quicksync you also need the following packages for Intel Gen 11 and above

1. `apk add intel-media-driver`
2. `apk add onevpl-intel-gpu --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing/`

For below 11th gen you probably need `intel-media-sdk` but I haven't confirmed or tested this.

To allow the container to use the GPU you need to do the following for incus and it should be similar for LXC/LXD or other container managers.

On the host run the following:
1. `getent group render | cut -d: -f3 `

Replace gid with the number that was presented from the first command.

2. `incus config device add "YOUR INSTANCE" gpu gpu gid=989`

You should now be able to run OME with
 `/usr/share/ovenmediaengine/OvenMediaEngine`

The Sever.xml config is located at:
`/usr/share/ovenmediaengine/conf/Server.xml`

I probably could have ran OME using docker within the Alpine container but I'm not a fan of docker for home use and prefer to avoid it when possible, I also decided on Alpine so it would have the smallest footprint possible on the system.

The total install size of the container with OME and QuickSync is 197MB

**Comment by saltomodules** [comment]
I can confirm the `intel-media-sdk` is needed to get Quicksync working on older Intel hardware, like Intel UHD Graphics 630.

Complete commandline:
 `apk add intel-media-driver intel-media-sdk onevpl-intel-gpu --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing/`

Also works with standalone ffmpeg build with `--enable-libvpl`.

**Comment by Blissingg** [comment]
If anyone is still interested in this and is having trouble with compiling v0.19 I will list the extra steps below.

Currently OME only works on Alpine 3.22 and lower as 3.23 and edge's ffmpeg versions are too new and now instead of features being deprecated they are fully gone. Could probably fix this with more editing but seemed long, ill just stick to 3.22.

Extra steps:
```
apk del spdlog-dev spdlog
cd /opt
git clone https://github.com/gabime/spdlog.git
cd spdlog && mkdir build && cd build
cmake .. && cmake --build .
make install

cd /opt/
git clone https://github.com/ggml-org/whisper.cpp.git
export PREFIX=/usr/share/whisper
export WHISPER_CUDA_ARCH="61;75;86"
cmake -B build -S . \
		-DCMAKE_BUILD_TYPE=Release \
		-DCMAKE_INSTALL_PREFIX=${PREFIX} \
		-DCMAKE_INSTALL_RPATH=${PREFIX}/lib \
		-DBUILD_SHARED_LIBS=ON \
		-DWHISPER_BUILD_EXAMPLES=OFF \
		-DWHISPER_BUILD_TESTS=OFF \
		-DWHISPER_BUILD_SERVER=OFF \
		-DGGML_CUDA=${WHISPER_CUDA} \
		-DCMAKE_CUDA_ARCHITECTURES=${WHISPER_CUDA_ARCH} && \
		cd build && \
		make -j 16 && \
		make install
```

You then have to edit the following:

`src/projects/base/ovlibrary/platform.h` on line 13 add the following: `#include <pthread.h>`

`src/projects/transcoder/codec/encoder/encoder_whisper.h`  line 11 change:
 `#include <whisper.h>`  to `#include </usr/share/whisper/include/whisper.h>`

That should be all that's needed to now compile and install v0.19 on Alpine 3.22 and lower.
